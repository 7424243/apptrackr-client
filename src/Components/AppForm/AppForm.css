.appform_container {
    display: flex;
    justify-content: center;
    flex-direction: column;
}

.appform_form {
    margin-left: 20px;
    max-width: 80%;
    text-align: left;
    padding-bottom: 40px;
}

.appform_input {
    margin-bottom: 10px;
}

.appform_header {
    margin-left: 20px;
    max-width: 80%;
    text-align: left;
    padding-top: 20px;
}