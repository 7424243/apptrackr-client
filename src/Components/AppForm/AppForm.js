import React, {Component} from 'react'
import ApptrackrContext from '../../ApptrackrContext'
import SquareButton from '../SquareButton/SquareButton'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSave } from '@fortawesome/free-solid-svg-icons'
import './AppForm.css'
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary'
import TokenService from '../../services/token-service'
import {isWebUri} from 'valid-url'
import config from '../../config'
import ValidationError from '../ValidationError/ValidationError'

class AppForm extends Component {

    state = {
        job_name: '',
        company_name: '',
        website_url: '',
        date_applied: '',
        contact_name: '',
        contact_email: '',
        contact_phone: '',
        interview_date: '',
        status: '',
        notes: '',
        user_id: '',
        error: null
    }

    static contextType = ApptrackrContext

    componentDidMount() {
        this.setState({user_id: TokenService.getUserIdFromToken()})
    }

    handleSubmitAdd = e => {
        e.preventDefault()
        let payload = Object.assign({}, this.state)
        fetch(`${config.API_ENDPOINT}/applications/`, {
            method: 'POST',
            body: JSON.stringify(payload),
            headers: {
                'content-type': 'application/json',
                'Authorization': `bearer ${TokenService.getAuthToken()}`
            }
        })
            .then(res => {
                if(!res.ok) {
                    return res.json().then(e => Promise.reject(e))
                }
                return res.json()
            })
            .then(data => {
                this.context.addApplication(data)
                this.props.history.push('/jobapps')
            })
            .catch(error => {
                this.setState({error})
                console.error({error})
            })
    }

    handleSubmitEdit = e => {
        console.log('submit for edit!')
    }

    addJobName = e => {
        this.setState({job_name: e.target.value})
    }

    addCompanyName = e => {
        this.setState({company_name: e.target.value})
    }

    addWebsiteUrl = e => {
        this.setState({website_url: e.target.value})
    }

    addDateApplied = e => {
        this.setState({date_applied: e.target.value})
    }

    addContactName = e => {
        this.setState({contact_name: e.target.value})
    }

    addContactPhone = e => {
        this.setState({contact_phone: e.target.value})
    }

    addContactEmail = e => {
        this.setState({contact_email: e.target.value})
    }

    addInterviewDate = e => {
        this.setState({interview_date: e.target.value})
    }

    addStatus = e => {
        this.setState({status: e.target.value})
    }

    addNotes = e => {
        this.setState({ntoes: e.target.value})
    }

    validateJobName() {
        const jobName = this.state.job_name.trim()
        if(jobName.length === 0) {
            return 'A job name is required'
        }
    }

    validateCompanyName() {
        const companyName = this.state.company_name.trim()
        if(companyName.length === 0) {
            return 'A company name is required'
        }
    }

    validateWebsiteUrl() {
        const url = this.state.website_url.trim()
        if(!isWebUri(url)) {
            return 'A valid URL is required'
        }
    }

    render() {

        const applicationId = parseInt(this.props.match.params.id)
        const {applications} = this.context
        const getApp = (applications, applicationId) => 
            applications.find(application => application.id === applicationId)
        const appDetails = getApp(applications, applicationId)

        return (
            <div className='appform_container'>
                <h2 className='appform_header'>{appDetails ? 'Edit Job Application' : 'New Job Application'}</h2>
                <ErrorBoundary>
                    <form 
                        className='appform_form'
                        onSubmit={appDetails ? this.handleSubmitEdit : this.handleSubmitAdd}
                    >
                        <section className='appform_input'>
                            <input 
                                type='text'
                                aria-label='job title'
                                name='job' 
                                defaultValue={appDetails ? appDetails.job_name : null} 
                                placeholder='Job Title'
                                required
                                onChange={this.addJobName}
                            />
                            {this.state.job_name && <ValidationError message={this.validateJobName()}/>}
                        </section>
                        <section className='appform_input'>
                            <input 
                                type='text'
                                aria-label='company name'
                                name='company' 
                                defaultValue={appDetails ? appDetails.company_name : null} 
                                placeholder='Company Name'
                                required
                                onChange={this.addCompanyName}
                            />
                            {this.state.company_name && <ValidationError message={this.validateCompanyName()}/>}
                        </section>
                        <section className='appform_input'>
                            <input 
                                type='url'
                                aria-label='website url'
                                name='website' 
                                defaultValue={appDetails ? appDetails.website : null} 
                                placeholder='Website'
                                onChange={this.addWebsiteUrl}
                            />
                            {this.state.website_url && <ValidationError message={this.validateWebsiteUrl()}/>} 
                        </section>
                        <section className='appform_input'>
                            <input 
                                type='text'
                                aria-label='date applied'
                                name='date-applied' 
                                defaultValue={appDetails ? appDetails.date_applied : null} 
                                placeholder='Date Applied'
                                onChange={this.addDateApplied}
                            />
                        </section>
                        <section className='appform_input'>
                            <input 
                                type='text'
                                aria-label='contact name'
                                name='contact' 
                                defaultValue={appDetails ? appDetails.contact_name : null} 
                                placeholder='Contact Name'
                                onChange={this.addContactName}
                            />
                        </section>
                        <section className='appform_input'>
                            <input 
                                type='text'
                                aria-label='contact email'
                                name='contact_email' 
                                defaultValue={appDetails ? appDetails.contact_email : null} 
                                placeholder='Contact Email'
                                onChange={this.addContactName}
                            />
                        </section>
                        <section className='appform_input'>
                            <input 
                                type='text'
                                aria-label='contact phone'
                                name='contact_phone' 
                                defaultValue={appDetails ? appDetails.contact_phone : null} 
                                placeholder='Contact Phone'
                                onChange={this.addContactPhone}
                            />
                        </section>
                        <section className='appform_input'>
                            <input 
                                type='text'
                                aria-label='interview date'
                                name='interview_date' 
                                defaultValue={appDetails ? appDetails.interview_date : null} 
                                placeholder='Interview Date'
                                onChange={this.addInterviewDate}
                            />
                        </section>
                        <section className='appform_input'>
                            <label htmlFor='status'>Status: </label>
                            <select required onChange={this.addStatus}>
                                <option name='status' value=''>Choose here</option>
                                <option name='status' value='Interested'>Interested</option>
                                <option name='status' value='Applied'>Applied</option>
                                <option name='status' value='Closed'>Closed</option>
                            </select>
                        </section>
                        <section className='appform_input appform_textarea'>
                            <textarea 
                                name='notes'
                                aria-label='additional notes'
                                defaultValue={appDetails ? appDetails.notes : null} 
                                placeholder='Additional Notes...'
                                onChange={this.addNotes}
                            />
                        </section>
                            <SquareButton type='submit'>
                                <FontAwesomeIcon icon={faSave}/>
                            </SquareButton>
                            {/* path={appDetails ? `/jobapp/${appDetails.id}` : '/jobapps'}
                            /> */}
                            <SquareButton
                                content={'X'}
                                path={appDetails ? `/jobapp/${appDetails.id}` : '/jobapps'}
                            />
                    </form>
                </ErrorBoundary>
            </div>
        )
    }
}

export default AppForm