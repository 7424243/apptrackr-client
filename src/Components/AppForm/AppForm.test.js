import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import {BrowserRouter} from 'react-router-dom'
import AppForm from './AppForm'

describe('AppForm Component', () => {
    
    const props = {
        match: {
            params: {
                id: '2'
            }
        },
        history: {
            push: jest.fn()
        }
    }

    //smoke test
    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDOM.render(<BrowserRouter><AppForm {...props}/></BrowserRouter>, div)
        ReactDOM.unmountComponentAtNode(div)
    })

    //snapshot test
    it('renders the UI as expected', () => {
        const tree = renderer
            .create(<BrowserRouter><AppForm {...props}/></BrowserRouter>)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
    
})