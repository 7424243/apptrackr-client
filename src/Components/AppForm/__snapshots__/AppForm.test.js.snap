// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppForm Component renders the UI as expected 1`] = `
<div
  className="appform_container"
>
  <h2
    className="appform_header"
  >
    New Job Application
  </h2>
  <form
    className="appform_form"
    onSubmit={[Function]}
  >
    <p>
      * Required
    </p>
    <section
      className="appform_input"
    >
      *
      <input
        aria-label="job title"
        className="input"
        defaultValue={null}
        name="job"
        onChange={[Function]}
        placeholder="Job Title"
        required={true}
        type="text"
      />
      
    </section>
    <section
      className="appform_input"
    >
      *
      <input
        aria-label="company name"
        className="input"
        defaultValue={null}
        name="company"
        onChange={[Function]}
        placeholder="Company Name"
        required={true}
        type="text"
      />
      
    </section>
    <section
      className="appform_input"
    >
      <input
        aria-label="website url"
        className="input"
        defaultValue={null}
        name="website"
        onChange={[Function]}
        placeholder="Website"
        type="url"
      />
      
    </section>
    <section
      className="appform_input"
    >
      <input
        aria-label="date applied ="
        className="input"
        defaultValue={null}
        name="date-applied"
        onChange={[Function]}
        placeholder="Date Applied mm/dd/yyyy"
        type="text"
      />
      
    </section>
    <section
      className="appform_input"
    >
      <input
        aria-label="contact name"
        className="input"
        defaultValue={null}
        name="contact"
        onChange={[Function]}
        placeholder="Contact Name"
        type="text"
      />
      
    </section>
    <section
      className="appform_input"
    >
      <input
        aria-label="contact email"
        className="input"
        defaultValue={null}
        name="contact_email"
        onChange={[Function]}
        placeholder="Contact Email"
        type="text"
      />
      
    </section>
    <section
      className="appform_input"
    >
      <input
        aria-label="contact phone"
        className="input"
        defaultValue={null}
        name="contact_phone"
        onChange={[Function]}
        placeholder="Contact Phone (xxx)xxx-xxxx"
        type="text"
      />
      
    </section>
    <section
      className="appform_input"
    >
      <input
        aria-label="interview date"
        className="input"
        defaultValue={null}
        name="interview_date"
        onChange={[Function]}
        placeholder="Interview Date mm/dd/yyyy"
        type="text"
      />
      
    </section>
    <section
      className="appform_input"
    >
      <label
        htmlFor="status"
      >
        * Status: 
      </label>
      <select
        defaultValue={null}
        onChange={[Function]}
        required={true}
      >
        <option
          name="status"
          value=""
        >
          Choose here
        </option>
        <option
          name="status"
          value="Interested"
        >
          Interested
        </option>
        <option
          name="status"
          value="Applied"
        >
          Applied
        </option>
        <option
          name="status"
          value="Closed"
        >
          Closed
        </option>
      </select>
    </section>
    <section
      className="appform_input appform_textarea"
    >
      <textarea
        aria-label="additional notes"
        className="textarea"
        defaultValue={null}
        name="notes"
        onChange={[Function]}
        placeholder="Additional Notes..."
      />
      
    </section>
    <div
      className="squarebutton_content"
    >
      <button
        className="squarebutton"
        type="submit"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-save fa-w-14 "
          data-icon="save"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </form>
</div>
`;
